runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -299,18 +304,33 @@ public void invalidateClientCache() {
      * @param addresses collection of addresses to read from.
      * @return A result to be cached
      */
-    public @Nonnull Map<Long, ILogData> cacheFetch(Iterable<Long> addresses) {
-        //turn the addresses into Set for now to satisfy signature requirement down the line
-        Set<Long> readAddresses = new TreeSet<>();
-        Iterator<Long> iterator = addresses.iterator();
-        while(iterator.hasNext()){
-            readAddresses.add(iterator.next());
+    public @Nonnull
+    Map<Long, ILogData> cacheFetch(Iterable<Long> addresses) {
+        Map<Long, ILogData> temp = new HashMap<>();
+
+        Iterable<List<Long>> batches = Iterables.partition(addresses, runtime.getBulkReadSize());
+
+        for (List<Long> batch : batches) {
+            //doesn't handle the case where some address have a different replication mode
+            temp.putAll(layoutHelper(l -> l.getReplicationMode(batch.iterator().next())
+                    .getReplicationProtocol(runtime)
+                    .readAll(l, batch)));
```

Comment:
No, we shouldn't fail the batch, indeed, we could actually just eliminate the batching code, as caffeine will automatically call loadAll again if it gets an incomplete result:

```
If the returned map doesn't contain all requested keys then the entries it does contain will be cached and getAll will return the partial results. If the returned map contains extra keys not present in keys then all returned entries will be cached, but only the entries for keys will be returned from getAll.
```